Add option to SabreLayout to specify starting layouts

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

The SabreLayout pass typically starts with each layout trial with a random starting layout. However, in some cases starting with a specific starting layout can result in better output quality than starting with a fully random layout. To use this feature an analysis pass can set a new `sabre_starting_layout` field in the property set before `SabreLayout` with a list of layout objects that will add additional trials using each layout as a starting layout.

### Details and comments

mtreinish committed:
Add option to SabreLayout to specify starting layouts

The SabreLayout pass typically starts with each layout trial with a
random starting layout. However, in some cases starting with a specific
starting layout can result in better output quality than starting with
a fully random layout. To use this feature an analysis pass can set a
new `sabre_starting_layout` field in the property set before
`SabreLayout` with a list of layout objects that will add additional
trials using each layout as a starting layout.

mtreinish committed:
Fix lint

jakelishman commented on crates/accelerate/src/sabre_layout.rs:
```diff
@@ -47,14 +50,19 @@ pub fn sabre_layout_and_routing(
+                    Some(&partial_layouts[num_layout_trials - index])
+                } else {
+                    None
+                };
```
Relatively minor style, but the API and logic might be clearer if we rename `num_layout_trials` to `num_random_trials`, then normalise both the random trials and `partial_layouts` into one single `partial_layouts` array at the start of the outer function - we can just use an empty vector for the full-random runs since they're a special case of "partial" that has _zero_ elements specified. Then we can unify the inner handling, and not need to worry about implicit indexing into either `num_layout_trials` or the `partial_layouts`, etc.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/10721/commits/51bbcd3bbcbfab26e65e12ca3fdd17deeb32f6cf

jakelishman commented on qiskit/transpiler/passes/layout/sabre_layout.py:
```diff
@@ -227,10 +241,16 @@ def run(self, dag):
+            inner_run = functools.partial(
+                self._inner_run, starting_layout=self.property_set["sabre_starting_layout"]
+            )
+
```
I'm assuming the idea of doing this rather than having it be a direct input to the `SabreLayout` is so a transpiler pass can look at the hardware / circuit and make a choice.  If we're starting to go down that route with passes (and we've already walked some of it with `ApplyLayout`, etc), perhaps we should consider formalising some sort of scope / namespacing in the property set?

mtreinish replied:
Yeah, typically the layout passes are dependent on the input circuit so using the property set was really the only clean way I could think of integrating it. The other option would be to take an optional callable input on `__init__` that is given a `DAGCircuit` and returns a layout. But the property set approach felt like a better fit for the transpiler. I think we can investigate adding more structure to the property set in a follow up. It's more than just apply layout, this same pattern also exists with the vf2 passes too because they can take in an optional error map from the property set too.

jakelishman replied:
Oh yeah I agree, the property set is the right way of feeding it in to a pass, if not only because it's what we already do (as you say).

jakelishman commented on qiskit/transpiler/passes/layout/sabre_layout.py:
```diff
@@ -69,6 +71,18 @@ class SabreLayout(TransformationPass):
+    list of :class:`.Layout` objects representing the starting layouts to use. If a
+    virtual qubit is missing from an :class:`.Layout` object in the list a random qubit
+    will be selected.
+
```
Related to the other comment in this file: since we're doing more of this now, I think it might be good to have a clearer way to document which property-set variables affect a pass's operation.  Something like
```rst
Properties Read
---------------

``sabre_starting_layout`` (``list[Layout]``)
    In addition to starting with a random initial :class:`.Layout`, ...

Properties Written
------------------
``layout`` (:class:`.Layout`)
    The chosen initial mapping of virtual to physical qubits, including the ancilla allocation.

``final_layout`` (:class:`.Layout`)
    A permutation of how swaps have been applied to the input qubits at the end of the circuit. 
```
(or whatever descriptions).

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/10721/commits/ebd74ad1cdd8ece706cb42269a2bac2340d04c13

jakelishman reviewed (commented):
I think this is a good change, and certainly useful for research into how different initial layouts affect the final layout / routing.

mtreinish committed:
Combine random layout and partial layouts

mtreinish committed:
Expand property set docs

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

jakelishman commented on crates/accelerate/src/sabre_layout.rs:
```diff
@@ -37,20 +39,24 @@ pub fn sabre_layout_and_routing(
         .collect();
     let dist = distance_matrix.as_array();
-    let res = if run_in_parallel && num_layout_trials > 1 {
+    let res = if run_in_parallel && (num_random_trials > 1 || !partial_layouts.is_empty()) {
```
I think this is `starting_layouts.len() > 1` now?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/10721/commits/f48d510c4fbcd877a98c4d98d3d2028faa5d20a9

jakelishman commented on crates/accelerate/src/sabre_layout.rs:
```diff
@@ -114,12 +123,33 @@ fn layout_trial(
     max_iterations: usize,
     num_swap_trials: usize,
     run_swap_in_parallel: bool,
+    starting_layout: &Vec<Option<usize>>,
```
Technically I think this could be `&[Option<usize>]` and we'd avoid a layer of pointer indirection.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/10721/commits/f588a4ff145e527c29021ded2915aeee9394666a

jakelishman commented on crates/accelerate/src/sabre_layout.rs:
```diff
@@ -114,12 +123,33 @@ fn layout_trial(
+        physical_qubits.shuffle(&mut rng);
+        physical_qubits
+    };
+
```
Minor, but I don't think we even need the `else` branch, right? The logic should work even if there's no `used_bits`.

mtreinish replied:
Yeah we don't need it, it would work fine without the if statement. But, I figured it was a bit more efficient to leave it like this. That being said I didn't measure it so I don't actually know how much overhead having a branch here vs doing creating 3 objects (one empty) and doing 2 iterations instead of one has. I can measure it to see if it's worth the extra code or not

jakelishman commented on releasenotes/notes/seed-sabre-with-layout-17d46e1a6f516b0e.yaml:
```diff
@@ -0,0 +1,38 @@
+    enables running an :class:`.AnalysisPass` prior to :class:`.SabreLayout`
+    which sets the ``"sabre_starting_layout"`` field in the property set
+    to provide the :class:`.SabreLayout` with additional starting layouts
+    to use in it's internal trials. For example, if you wanted to run
```
```suggestion
    to use in its internal trials. For example, if you wanted to run
```

jakelishman commented on qiskit/transpiler/passes/layout/sabre_layout.py:
```diff
@@ -69,6 +71,34 @@ class SabreLayout(TransformationPass):
+
+    ``final_layout`` (:class:`.Layout`)
+        A permutation of how swaps have been applied to the input qubits at the end of the circuit.
+
```
Let's follow up on this in a different PR to make our other transpiler passes have a section like this too.

jakelishman replied:
Follow-up in #10727.

jakelishman reviewed (dismissed):
I think most of my stuff is just nitpicks, so feel free to ignore.  I'm fine to merge as is, but I'd want Sasha to approve as well before, since this is intended for work he's doing as well.

mtreinish committed:
Simplify branching logic for parallelism

mtreinish committed:
Use a slice for starting_layout in a trial

mtreinish committed:
Merge remote-tracking branch 'origin/main' into partial-layout-interface

mtreinish committed:
Update releasenotes/notes/seed-sabre-with-layout-17d46e1a6f516b0e.yaml

Co-authored-by: Jake Lishman <jake@binhbar.com>

jakelishman reviewed (dismissed):
(same comment about waiting for Sasha)

mtreinish committed:
Merge remote-tracking branch 'origin/main' into partial-layout-interface

mtreinish committed:
Merge remote-tracking branch 'origin/main' into partial-layout-interface

alexanderivrii commented on qiskit/transpiler/passes/layout/sabre_layout.py:
```diff
@@ -312,7 +346,7 @@ def run(self, dag):
         return mapped_dag
 
-    def _inner_run(self, dag, coupling_map):
+    def _inner_run(self, dag, coupling_map, starting_layout=None):
```
Would it be slightly clearer to call this `starting_layouts` instead of `starting_layout`?

alexanderivrii reviewed (dismissed):
Thanks, this is all very nice! I especially like the fact that one can write multiple "guess the initial layout"  passes, each pass adding its own layout(s) to `property_set["sabre_starting_layout"]`, and then `SabreLayout` would examine all of these in addition to the prescribed number of random layouts. 

The only super minor question if you think that these initial layouts may eventually make sense also outside of sabre; if so, maybe let's call these `"starting_layout"` and not `"sabre_starting_layout"`?  Also, should it be `"sabre_starting_layouts"` to emphasize that it's a list; the same for the new option in `SabreLayout`'s `_inner_run`. But please feel free to ignore this.

mtreinish commented:
> The only super minor question if you think that these initial layouts may eventually make sense also outside of sabre; if so, maybe let's call these "starting_layout" and not "sabre_starting_layout"? Also, should it be "sabre_starting_layouts" to emphasize that it's a list; the same for the new option in SabreLayout's _inner_run. But please feel free to ignore this.

I'm not sure there is an application for this outside of sabre layout. Our other in tree layout passes don't really work with a starting layout the way sabre does. I'm having a hard time coming up with another use case where we'd want to feed an unapplied layout as an input to a pass like this does (even `VF2PostLayout` needs the layout to be applied and routed to work). I think namespacing it to sabre makes sense to clearly indicate that.

As for the variable names yeah you're correct they should probably be plural. I'll update the properties and arguments to be plural `starting_layouts` instead of singular `starting_layout` as that is more descriptive as it's a list.

mtreinish committed:
Make starting layout variable names plural

The starting layout is a list and can have more than one entry. To make
this clearer, this commit renames starting_layout -> starting_layouts
and sabre_starting_layout -> sabre_starting_layouts.

mtreinish committed:
Merge branch 'main' into partial-layout-interface

jakelishman commented on releasenotes/notes/seed-sabre-with-layout-17d46e1a6f516b0e.yaml:
```diff
@@ -0,0 +1,38 @@
+
+          def run(self, dag):
+              self.dense_pass.run(dag)
+              self.property_set["sabre_starting_layout"] = [self.dense_pass.property_set["layout"]]
```
```suggestion
              self.property_set["sabre_starting_layouts"] = [self.dense_pass.property_set["layout"]]
```

jakelishman commented on test/python/transpiler/test_sabre_layout.py:
```diff
@@ -218,6 +253,18 @@ def test_layout_many_search_trials(self):
+    def run(self, dag):
+        self.dense_pass.run(dag)
+        self.property_set["sabre_starting_layout"] = [self.dense_pass.property_set["layout"]]
+
```
Given that this test isn't failing while it's still setting the incorrect property-set entry (no terminal "s"), I'm a bit concerned that the test isn't doing its job haha.

mtreinish replied:
Hah, well it was too tricky to come up with an example where using dense layout actually was the selected layout. The full random was almost always a better starting point, so the test just really to exercise the code paths but couldn't enforce they were used easily because we lose it to rust space at the point wed want to assert anything.

mtreinish committed:
Update releasenotes/notes/seed-sabre-with-layout-17d46e1a6f516b0e.yaml

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Merge branch 'main' into partial-layout-interface

mtreinish committed:
Update property set key in test

jakelishman commented on test/python/transpiler/test_sabre_layout.py:
```diff
@@ -94,6 +94,41 @@ def test_6q_circuit_20q_coupling(self):
+        )
+        pm.run(circuit)
+        layout = pm.property_set["layout"]
+        self.assertEqual([layout[q] for q in circuit.qubits], [7, 8, 12, 6, 11, 13])
```
Maybe we should set this to make zero random trials, so we can at least assert that the given initial layout is being tried?  Given we know that the output currently doesn't depend on the input map, I'm a bit nervous about merging a test that would have the same result even if the code paths it's purportedly testing are completely broken.

mtreinish replied:
Oh I hadn't thought of setting 0 layout trials, I can give that a try.

mtreinish replied:
Using 0 trials works, if the property set key is incorrect it raises a PanicException, which while not the most ergonomic at least points to an issue in the code here. I made the change in: https://github.com/Qiskit/qiskit/pull/10721/commits/13e49ab4b08c8d25947736e7398a6caade310fc7

mtreinish committed:
Update test to only use partial layout

mtreinish committed:
Merge branch 'main' into partial-layout-interface

jakelishman reviewed (approved):
Thanks for all the changes - now that Sasha's said he's happy too I'm good to merge.
